File: 2023/01/01.py
'''
# 1
sum = 0
with open('1.txt') as f:
    for line in f:
        for i in line:
            if i.isdigit():
                sum += int(i)*10
                break
        for i in line[::-1]:
            if i.isdigit():
                sum += int(i)
                break
print(sum)
'''

# 2
sum = 0
digit = None
with open('1.txt') as f:
    for line in f:
        found_first_digit = False
        found_digit = False
        for idx, i in enumerate(line):
            if i.isdigit():
                digit = i
                found_digit = True
            elif line[idx:idx+3] == 'one':
                digit = 1
                found_digit = True
            elif line[idx:idx+3] == 'two':
                digit = 2
                found_digit = True
            elif line[idx:idx+5] == 'three':
                digit = 3
                found_digit = True
            elif line[idx:idx+4] == 'four':
                digit = 4
                found_digit = True
            elif line[idx:idx+4] == 'five':
                digit = 5
                found_digit = True
            elif line[idx:idx+3] == 'six':
                digit = 6
                found_digit = True
            elif line[idx:idx+5] == 'seven':
                digit = 7
                found_digit = True
            elif line[idx:idx+5] == 'eight':
                digit = 8
                found_digit = True
            elif line[idx:idx+4] == 'nine':
                digit = 9
                found_digit = True

            if found_digit and not found_first_digit:
                sum += int(digit)*10
                found_first_digit = True
        else:
            sum += int(digit)
print(sum)

File: 2023/01/01.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
	"unicode"
)

func p1(filename string) {
	file, err := os.Open(filename)
	if err != nil {
		fmt.Println(err.Error())
		return
	}
	defer file.Close()

	res := 0
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		var digits[]rune
		for _, ch := range scanner.Text() {
			if unicode.IsDigit(ch) {
				digits = append(digits, ch)
			}
		}
		res += int(digits[0] - '0') * 10 + int(digits[len(digits)-1] - '0')
	}
	fmt.Println(res)
}


func p2(filename string) {
	f, _ := os.Open(filename)
	defer f.Close()

	m := map[string]string{
		"one": "o1e",
		"two": "t2o",
		"three": "t3e",
		"four": "f4r",
		"five": "f5e",
		"six": "s6x",
		"seven": "s7n",
		"eight": "e8t",
		"nine": "n9e",
	}

	res := 0
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		for target, new := range m {
			line = strings.Replace(line, target, new, -1)
		}

		digits := regexp.MustCompile(`\d`).FindAllString(line, -1)
		d1, _ := strconv.Atoi(digits[0])
		dn, _ := strconv.Atoi(digits[len(digits)-1])
		res += d1 * 10 + dn
	}
	fmt.Println(res)
}

func main() {
	// input := "example.txt"
	input := "input.txt"
	// p1(input)
	p2(input)
}
